helper/startup.go

```diff
@@ -32,59 +25,54 @@ func InteractiveHandshake(settings *crestmarket.OAuthSettings, store *FileTokenS
 		return nil, err
 	}
-	t, err := f.NewTransportFromCode(code)
+
+	// TODO: Explore how to get Mediate's retrying transport layer adapted
+	// into the Context-to-http.Client lookup table before making the
+	// exchange.
+	// Would be nice if oauth2 exported registerContextClientFunc() ...bummer
+	freshToken, err := config.Exchange(oauth2.NoContext, code)
 	if err != nil {
 		log.Fatal(err)
 		return nil, err
 	}
-	return t, nil
-}
-
-// BackgroundStartup tries to start with the current token
-// or fails with a Fatal if the token can't be read or is expired.
-func BackgroundStartup(tokenFile string, settings *crestmarket.OAuthSettings) (http.RoundTripper, error) {
-	store := FileTokenStore{Filename: tokenFile}
-
-	base, err := crestmarket.NewOAuthOptions(settings)
-	t, err := base.NewTransportFromTokenStore(&store)
-	if err != nil {
-		log.Fatal("Token refresh has failed")
-	}
-	_, err = t.CheckAndRefreshToken()
-	if err != nil || t.Token().Expired() {
-		log.Fatal("Token is expired and refresh has failed.")
-	}
-	store.WriteToken(t.Token())
-	return t, nil
 
+	return freshToken, nil
 }
 
 // InteractiveStartup performs a console interactive handshake
 // or a simple refresh of tokens and stores
 // tokens gathered in a file called token.json
 func InteractiveStartup(tokenFile string, settings *crestmarket.OAuthSettings) (http.RoundTripper, error) {
-	store := FileTokenStore{Filename: tokenFile}
+	config := crestmarket.NewOAuthConfig(settings)
+	source := NewFileTokenSource(tokenFile)
 
-	base, err := crestmarket.NewOAuthOptions(settings)
-	t, err := base.NewTransportFromTokenStore(&store)
-	if err != nil {
-		log.Println("Token refresh has failed, requesting new authorization interactively")
-		t, err = InteractiveHandshake(settings, &store)
+	// Try the new FileTokenSource, if it doesn't produce a valid token, force
+	// the user into the interactive prompt.
+	if token, err := source.Token(); err != nil || token.AccessToken == "" {
+		log.Println("Token is not valid, requesting new authorization interactively")
+		source.CachedToken, err = InteractiveHandshake(config)
 		if err != nil {
 			log.Println("Can't really continue, auth has failed.")
 			return nil, err
 		}
+		// TODO: Update the token whenever it gets refreshed by OAuth2.
+		// Context: We only write to the file if the user goes through
+		// interactive handshake. If the token file contains a valid but
+		// expired token, OAuth2 will (internally) go refresh the token on
+		// its first RoundTrip() with the server. We never end up writing
+		// the refreshed token back to the file.
+		// Perhaps put this functionality into a time.NewTicker?
+		source.WriteTokenToFile()
 	}
-	_, err = t.CheckAndRefreshToken()
-	if err != nil || t.Token().Expired() {
-		log.Fatal("Token is expired and refresh has failed.")
+
+	transport := &oauth2.Transport{
+		// Use Config.TokenSource to make use of oauth2.tokenRefresher{...}
+		Source: config.TokenSource(oauth2.NoContext, source.CachedToken),
```

Comment:
This line fixed the issue where tokens weren't getting refreshed. Looking at the oauth2 api, this appears to be the only means of using oauth2.tokenRefresher. However, we no longer pass the FileTokenSource into the oauth2.Transport... thus removing the only reason FileTokenSource implemented the TokenSource interface.